library/alloc/src/rc.rs

```diff
@@ -1892,15 +1879,17 @@ impl<T> Weak<T> {
     pub fn as_ptr(&self) -> *const T {
         let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);
 
-        // SAFETY: we must offset the pointer manually, and said pointer may be
-        // a dangling weak (usize::MAX) if T is sized. data_offset is safe to call,
-        // because we know that a pointer to unsized T was derived from a real
-        // unsized T, as dangling weaks are only created for sized T. wrapping_offset
-        // is used so that we can use the same code path for the non-dangling
-        // unsized case and the potentially dangling sized case.
-        unsafe {
-            let offset = data_offset(ptr as *mut T);
-            set_data_ptr(ptr as *mut T, (ptr as *mut u8).wrapping_offset(offset))
+        if is_dangling(self.ptr) {
+            // If the pointer is dangling, we return a null pointer as the dangling sentinel.
+            // We can't return the usize::MAX sentinel, as that could valid if T is ZST.
```

Comment:
Thinking about it again, I don't think I agree with this argument... due to the alignment requirement of `RcBox` and its `repr(C)` layout, `&raw mut (*ptr).value` will always be ptr-aligned, so it can never be `usize::MAX`. Am I missing something?

Reply:
No, that's correct; I missed that reasoning originally. We could just return the `usize::MAX` sentinel directly. `is_dangling` currently takes `NonNull` rather than a raw pointer, so there'd be a few more changes than just using `is_dangling` rather than `is_null` in `from_raw`, but it would work.

I did [a quick test on godbolt](https://rust.godbolt.org/z/Tbc4eq) and using `is_dangling` is (as intuitively) less work, so I'll switch over to just returning that sentinel directly.

<details><summary>ASM</summary>

```asm
; returning null as the outside sentinel
example::as_ptr_1:
        mov     rdx, rsi
        lea     rcx, [rdi + 16]
        xor     eax, eax
        cmp     rdi, -1
        cmovne  rax, rcx
        ret

example::from_ptr_1:
        mov     rdx, rsi
        lea     rcx, [rdi - 16]
        xor     eax, eax
        cmp     rdi, 1
        sbb     rax, rax
        or      rax, rcx
        ret

; returning MAX as the outside sentinel
example::as_ptr_2:
        mov     rdx, rsi
        lea     rax, [rdi + 16]
        cmp     rdi, -1
        cmove   rax, rdi
        ret

example::from_ptr_2:
        mov     rdx, rsi
        lea     rax, [rdi - 16]
        cmp     rdi, -1
        cmove   rax, rdi
        ret
```
</details>